### ISSUE TITLE
BeEquivalentTo and NotBeEquivalentTo consistency

### ISSUE THREAD

**Issue by djmeijer** [fluentassertions/fluentassertions#1036]
### Description

I may have found some weird cases when using the BeEquivalentTo and NotBeEquivalentTo functions. To put it short:

1. Using compile time type to check equivalence seems a little bit weird.
2. BeEquivalentTo is not mutually exclusive with NotBeEquivalentTo on IEnumerable.
3. NotBeEquivalentTo is not implemented on an object where BeEquivalentTo is.
4. BeEquivalentTo on an object is not commutative.

I like to hear some other views and opinions on this.

### Complete minimal example reproducing the issue

The test cases 1 to 6 represent my line of thought.

```csharp
using System.Collections.Generic;
using FluentAssertions;
using FluentAssertions.Execution;
using NUnit.Framework;

namespace FluentAssertionTests
{
  public abstract class Car
  {
    protected Car(string licensePlate)
    {
      LicensePlate = licensePlate;
    }

    public string LicensePlate { get; }
  }

  public class Peugeot : Car
  {
    public Peugeot(string licensePlate, string peugeotNumber)
      : base(licensePlate)
    {
      PeugeotNumber = peugeotNumber;
    }

    public string PeugeotNumber { get; }
  }

  [TestFixture]
  public class Tests
  {
    private Car _expected;
    private IEnumerable<Car> _expectedIEnumerable;
    private Car _actual;
    private IEnumerable<Car> _actualIEnumerable;

    /// <summary>
    ///   I expect all the statements in this scope to fail, but they all succeed.
    /// </summary>
    [Test]
    public void TestCase1()
    {
      using(new AssertionScope())
      {
        _expectedIEnumerable.Should().BeEquivalentTo(_actualIEnumerable);
        new[] { _expected }.Should().BeEquivalentTo(_actualIEnumerable);
        _expected.Should().BeEquivalentTo(_actual);
      }
    }

    /// <summary>
    ///   All the statements fail, as I expect.
    /// </summary>
    [Test]
    public void TestCase2()
    {
      using(new AssertionScope())
      {
        new[] { _expected }.Should().BeEquivalentTo(new[] { _actual });
        _expectedIEnumerable.Should().BeEquivalentTo(new[] { _actual });
      }
    }

    /// <summary>
    ///   In order to let TestCase1 fail, the RespectingRuntimeTypes option has to be enabled. Can this be the default instead
    ///   of compile time type?
    /// </summary>
    [Test]
    public void TestCase3()
    {
      using(new AssertionScope())
      {
        _expectedIEnumerable.Should().BeEquivalentTo(_actualIEnumerable, o => o.RespectingRuntimeTypes());
        new[] { _expected }.Should().BeEquivalentTo(_actualIEnumerable, o => o.RespectingRuntimeTypes());
        _expected.Should().BeEquivalentTo(_actual, o => o.RespectingRuntimeTypes());
      }
    }

    /// <summary>
    ///   I expect BeEquivalentTo and NotBeEquivalentTo to be mutually exclusive. But this test succeeds.
    /// </summary>
    [Test]
    public void TestCase4()
    {
      using(new AssertionScope())
      {
        _expectedIEnumerable.Should().BeEquivalentTo(_actualIEnumerable);
        _expectedIEnumerable.Should().NotBeEquivalentTo(_actualIEnumerable);
      }
    }

    /// <summary>
    ///   There is no ObjectAssertions NotBeEquivalentTo method. Can this be added?
    /// </summary>
    [Test]
    public void TestCase5()
    {
      using(new AssertionScope())
      {
        _expected.Should().BeEquivalentTo(_actual);
        //_expected.Should().NotBeEquivalentTo(_actual);
      }
    }

    /// <summary>
    ///   Here I expect both statements to fail, only the second one fails.
    /// </summary>
    [Test]
    public void TestCase6()
    {
      using(new AssertionScope())
      {
        new Peugeot("AA-AA-AA", "123").Should().BeEquivalentTo(_actual);
        _expected.Should().BeEquivalentTo(new Peugeot("AA-AA-AA", "456"));
      }
    }

    [OneTimeSetUp]
    public void OneTimeSetUp()
    {
      _expected = new Peugeot("AA-AA-AA", "123");
      _expectedIEnumerable = new[] { _expected };
      _actual = new Peugeot("AA-AA-AA", "456");
      _actualIEnumerable = new[] { _actual };
    }
  }
}
```

### Expected behavior:

1. Using runtime type to check equivalence as default. This has the most information value. Create the inverse option to use the compile time type when needed. Potentially a lot of tests will break.
2. Make BeEquivalentTo mutually exclusive with NotBeEquivalentTo given the same parameters. Either things are equivalent, or they are not. Exclusive or. Law of excluded middle.
3. Because BeEquivalentTo is implemented on objects, so will NotBeEquivalentTo.
4. BeEquivalentTo is commutative. So left.BeEquivalentTo(right) will yield the same result as right.BeEquivalentTo (left). The same applies to NotBeEquivalentTo.

### Actual behavior:

See the 6 test cases. Please read them top down and the actual behavior will be clear.

### Versions

Using version 5.6.0 of Fluent Assertions and .NETCoreApp 2.2.

**Comment by dennisdoomen** [comment]
Thanks for taking the time to come up with all those well-described examples. Let me answer your suggestions one by one.

> Using runtime type to check equivalence as default.

This is a matter of taste IMO. You're the first one ever to suggest this alternative. If we would change the default, I'm pretty sure people will start to complain as well. We chose to honor the compile-time type as much as possible. 

> Make BeEquivalentTo mutually exclusive with NotBeEquivalentTo given the same parameters.

This is weird indeed. I'll need to investigate. 

> Because BeEquivalentTo is implemented on objects, so will NotBeEquivalentTo.

This is just missing functionality that nobody has asked for

> So left.BeEquivalentTo(right) will yield the same result as right.BeEquivalentTo (left).

This is by design. If the `left` object has more properties or fields than the `right` object, the first assertion  will succeed. However, if it would be other way around, it would fail. This changed in 5.0 and allows you to compare any object against an anonymous type that only includes the properties you care about.

**Comment by jnyrup** [comment]
### testcase1
`RespectingRuntimeTypes` is necessary, as the compile time type of the expectation drives the comparison.

### testcase2
* both `BeEquivalentTo` resolves to `BeEquivalentTo(params object[] expectations)`

`params object[]` uses runtime types.

### testcase4
* `BeEquivalentTo` resolves to `BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ...`
* `NotBeEquivalentTo` resolves to `NotBeEquivalentTo(IEnumerable unexpected, ...`
	
`IEnumerable<TExpectation>` uses compile time types
`IEnumerable` uses runtime types
	
### testcase5
* `BeEquivalentTo<Car>(Car expectation)`
* `BeEquivalentTo<Peugeot>(Peugeot expectation)`

The different behavior is due to the different compile time type of the `expectation`.

I agree that the `IEnumerable<T>` vs `IEnumerable`/`params object[]` causes inconsistencies and headaches.

**Comment by bacar** [comment]
> This is a matter of taste IMO

I don't think it's just a matter of taste.
Respecting the runtime type when testing equivalence is _more conservative_ - that is, it is more likely to fail if you accidentally misconfigure it. If you need the less restrictive check, you will see the assertion fail and modify your test accordingly. The way things are at the moment, it is too easy to unexpectedly test too weakly - I suspect our and many other codebases are rife with such errors, because the library does not encourage you to discover such problems.

However, the more conservative option for `NotBeEquivalentTo()` is to respect the compile time type, so it's not possible to meet the dual goals of doing a more conservative check and having `BeEquivalentTo` 
 be consistent with `NotBeEquivalentTo` :-/

**Comment by dennisdoomen** [comment]
> Respecting the runtime type when testing equivalence is more conservative - that is, it is more likely to fail if you accidentally misconfigure it. 

This is true, but over the years, we've also had bug reports that people where comparing two object graphs through an interface and did not expect the properties of the run-time type to be included.
